### PR TITLE
Faster Pkg.Publish()

### PR DESCRIPTION
**Pull request by carlobaldassi**
First step at dealing with #10323 issues. Restricts checking to potentially affected packages.
Unfortunately, it doesn't help as much as I hoped, since the (undirected) dependency graph basically percolates through Compat. Needs more smartness; it's a start though.

@StefanKarpinski: I moved the check later in `Pkg.publish`, I think it should be ok, but please check.

### CODE REVIEW

**Comment by timholy** [comment]
Thanks for tackling this! What are the pre/post performance numbers like?

**Comment by carlobaldassi** [comment]
> What are the pre/post performance numbers like?

Alas, almost equal for "connected" packages (e.g. for HDF5 you'd save 2s at most). For "singleton" packages instead (e.g. MNIST, Gaston) `check_metadata` goes down from 108s to less than 1s on my machine.

**Comment by carlobaldassi** [comment]
So now I quickly applied some of Tim's ideas from #10323. I still want to try other ways to do the same, since the code is really ugly, but it works, and it already gets from 108s to 78s on my machine (and from 29% to 5% gc time). I think I'll merge when it turns green, and keep working on it from there.

File: base/pkg/resolve/versionweight.jl [hunk 1]
```diff
@@ -109,25 +111,41 @@ function VWPreBuild(ispre::Bool, desc::(Union(Int,ASCIIString)...))
...
+Base.typemin(::Type{VWPreBuild}) = _vwprebuild_min
 
-Base.(:-)(a::VWPreBuild, b::VWPreBuild) = VWPreBuild(a.nonempty-b.nonempty, a.w-b.w)
-Base.(:+)(a::VWPreBuild, b::VWPreBuild) = VWPreBuild(a.nonempty+b.nonempty, a.w+b.w)
+function Base.(:-)(a::VWPreBuild, b::VWPreBuild)
+    b === _vwprebuild_zero && return a
+    a === _vwprebuild_zero && return -b
+    VWPreBuild(a.nonempty-b.nonempty, a.w-b.w)
+end
+function Base.(:+)(a::VWPreBuild, b::VWPreBuild)
```
Comment (tkelman) [inline base/pkg/resolve/versionweight.jl]: trailing whitespace here

**Comment by carlobaldassi** [comment]
Well, it's 30s now...
I tried more tricks but they didn't help much.

**Comment by carlobaldassi** [comment]
Build failure is appveyor timing out.

**Comment by carlobaldassi** [comment]
@timholy: thanks a lot for the profiling and the pointers. I hope this allows to use `publish` again for now. (And I still hope to improve this further somehow.)

**Comment by timholy** [comment]
The thanks are owed to you! I'm really looking forward to trying it.

**Pull request merged**